x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DocVector.java

```diff
@@ -127,45 +128,59 @@ private void buildShardSegmentDocMapIfMissing() {
             return;
         }
 
-        int[] forwards = shardSegmentDocMapForwards = new int[shards.getPositionCount()];
-        for (int p = 0; p < forwards.length; p++) {
-            forwards[p] = p;
-        }
-        new IntroSorter() {
-            int pivot;
-
-            @Override
-            protected void setPivot(int i) {
-                pivot = forwards[i];
+        boolean success = false;
+        long estimatedSize = sizeOfSegmentDocMap();
+        blockFactory().adjustBreaker(estimatedSize, true);
+        try {
+            int[] forwards = shardSegmentDocMapForwards = new int[shards.getPositionCount()];
+            for (int p = 0; p < forwards.length; p++) {
+                forwards[p] = p;
             }
+            new IntroSorter() {
+                int pivot;
 
-            @Override
-            protected int comparePivot(int j) {
-                int cmp = Integer.compare(shards.getInt(pivot), shards.getInt(forwards[j]));
-                if (cmp != 0) {
-                    return cmp;
+                @Override
+                protected void setPivot(int i) {
+                    pivot = forwards[i];
                 }
-                cmp = Integer.compare(segments.getInt(pivot), segments.getInt(forwards[j]));
-                if (cmp != 0) {
-                    return cmp;
+
+                @Override
+                protected int comparePivot(int j) {
+                    int cmp = Integer.compare(shards.getInt(pivot), shards.getInt(forwards[j]));
+                    if (cmp != 0) {
+                        return cmp;
+                    }
+                    cmp = Integer.compare(segments.getInt(pivot), segments.getInt(forwards[j]));
+                    if (cmp != 0) {
+                        return cmp;
+                    }
+                    return Integer.compare(docs.getInt(pivot), docs.getInt(forwards[j]));
                 }
-                return Integer.compare(docs.getInt(pivot), docs.getInt(forwards[j]));
-            }
 
-            @Override
-            protected void swap(int i, int j) {
-                int tmp = forwards[i];
-                forwards[i] = forwards[j];
-                forwards[j] = tmp;
-            }
-        }.sort(0, forwards.length);
+                @Override
+                protected void swap(int i, int j) {
+                    int tmp = forwards[i];
+                    forwards[i] = forwards[j];
+                    forwards[j] = tmp;
+                }
+            }.sort(0, forwards.length);
 
-        int[] backwards = shardSegmentDocMapBackwards = new int[forwards.length];
-        for (int p = 0; p < forwards.length; p++) {
-            backwards[forwards[p]] = p;
+            int[] backwards = shardSegmentDocMapBackwards = new int[forwards.length];
+            for (int p = 0; p < forwards.length; p++) {
+                backwards[forwards[p]] = p;
+            }
+            success = true;
```

Comment:
I think we need to move the assignment of `shardSegmentDocMapForwards` and `shardSegmentDocMapBackwards` after `success=true`. Otherwise, we might call adjustBreaker twice: once in the finally block and once in the closeInternal method.